Secrets design proposal (#1989)
Azure/azure-service-operator: docs/hugo/content/design/secrets.md

   issue an instruction to us to perform an action to roll the secret, which doesn't fit as well into the goal-seeking paradigm Kubernetes prefers.
2. It takes control away from the user. What if the secret the operator generates doesn't comply with a particular organizations complexity requirements?
3. It requires that we do a _very_ good job generating cryptographically secure passwords/secrets. This could easily become a can of worms.
4. It hinders adoption of existing resources. If the operator expects to always generate the secret for a SQL DB, but the user wants to import a SQL DB 
   they've already created through some other mechanism, then they are at an impasse as there is no (easy) way for them to provide ASO with the secret.
5. It doesn't work well with GitOps. A number of customers have expressed the desire to move resources between namespaces. In theory this is easy - 
   just create the exact same resources in a different namespace (pointing at the same Azure resources), mark the old resources as `skip-deletion`, and 

>    delete the old resources. If the definition of the secret isn't part of the users GitOps flow this becomes more difficult as the first namespace has 

Author: Clarified

   the secret (which was created by ASO) and that secret must be cloned manually to the second namespace, otherwise the creation of the resource in the 
   second namespace will attempt to generate a new secret.

We could investigate allowing both automatically generated secrets (the default) and user specified secrets (opt-in). That avenue results in the operator 
shouldering all of the complexity burden of both though. Given ASO's approach as a low level toolkit providing the ability to create Azure resources, at least 
at this time we should avoid the added complexity of generating user secrets.

> 

Author: Done

### Prefer Managed Identity

Managing secrets is hard. We should prefer AAD Authentication and Managed Identity where possible.

What this means in practice is that secrets should not be retrieved **unless the user has specifically asked for them to be**.

This has a number of advantages:
1. If the user _is_ using Managed Identity, they don't have a secret leak waiting to happen created in their namespace which they didn't ask about 
   (and might not even know about).

> 2. If the user _is_ using Managed Identity, many services support ways to instruct the `ListKeys` (or equivalent) API to error when called, as a way to

Author: > Are you just saying we should only call ListKeys at the user's behest (controlled by some for-operator part of the spec), so that we can interpret that error as fatal?

Yes. I've restructured this sentence quite a bit to make the important part come first.

   force all callers to use an AAD Identity. If the operator is not calling those APIs unless instructed, we can make their failure fatal - the user must
   remove the bit of the `Spec` instructing us to call that API for their resource to reconcile successfully.

We'll dig more into how we might accomplish this a bit later.

### Work well with GitOps

Resources should strive to play well with GitOps. This means that when a new resource is deployed it is capable of adopting an already existing ARM resource
if one exists. This also must apply to secrets. As mentioned above one of the reasons that 
[we do not want to be generating secrets on the users behalf](#aso-should-not-be-generating-or-managing-secrets-on-the-users-behalf) is that it makes redeploy
and resource adoption hard because the user has no configuration representing the generated secret.

## Kinds of secrets

**User provided secrets**: Secrets provided by the user at resource creation time, and never returned by Azure.

> **Azure generated secrets**: Secrets created by Azure, and returned in a special `GetMeTheSecrets` API call.

Author: Good idea - fixed.

> | CRD                        | User provided secrets | Azure generated secrets | AAD/Managed Identity Support | Notes                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |

Author: I've updated this to have a clearer title as well, which I think should resolve this. It's now under a heading `Sample resources that have secrets`

| CosmosDB DatabaseAccount   | ❌                     | ✔️                       | ✔️                            | [List Keys API](https://docs.microsoft.com/en-us/rest/api/cosmos-db-resource-provider/2021-04-15/database-accounts/list-keys), [List Read Only Keys](https://docs.microsoft.com/en-us/rest/api/cosmos-db-resource-provider/2021-04-15/database-accounts/list-read-only-keys) and [Regenerate Key API](https://docs.microsoft.com/en-us/rest/api/cosmos-db-resource-provider/2021-04-15/database-accounts/regenerate-key). For AAD+RBAC (supported by SQL only?), see [Disabling Local Auth](https://docs.microsoft.com/en-us/azure/cosmos-db/how-to-setup-rbac#disable-local-auth), [Create Role Assignment API](https://docs.microsoft.com/en-us/rest/api/cosmos-db-resource-provider/2021-04-15/sql-resources/create-update-sql-role-assignment), [Create Role Definition API](https://docs.microsoft.com/en-us/rest/api/cosmos-db-resource-provider/2021-04-15/sql-resources/create-update-sql-role-definition). [Built-in Role Definitions](https://docs.microsoft.com/en-us/azure/cosmos-db/how-to-setup-rbac#built-in-role-definitions). |

> | EventHubAuthorizationRules | ❌                     | ✔️                       | ❌                            | [List Keys API](https://docs.microsoft.com/en-us/rest/api/eventhub/stable/authorization-rules-event-hubs/list-keys). There are default authorization rules created, such as `RootManageSharedAccessKey`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |

Author: Fixed.

<!-- TODO: ServiceBus namespace? -->

### Other kinds of secrets in Azure:

1. Get once Azure generated secrets: These are secrets created by Azure and _only returned once_, usually as the response to a POST.
   These don't fit cleanly into the table above because they are a POST action on a parent resource these are not in fact resources themselves.
   1. Application Insights Component APIKey: This is a POST to the Component/ApiKey URL.
   2. KeyVault Key: [Create Key API](https://docs.microsoft.com/en-us/rest/api/keyvault/create-key/create-key).
2. "Secrets" created by Azure, and returned in the GET: These "secrets" are not really secrets per-se, but ASOv1 classifies them as secrets.
   1. `ApplicationInsights` `InstrumentationKey` or `ConnectionString`? 
      See [here](https://docs.microsoft.com/en-us/rest/api/application-insights/components/create-or-update#applicationinsightscomponent).
3. Short-lived "tokens".
   1. `StorageAccount` SAS
   2. `CosmosDB` ResourceToken.

## Other Operators

A quick look at what other operators are doing with regard to secrets.

### Crossplane

**Azure generated secrets**:

> A [Connection secret](https://doc.crds.dev/github.com/crossplane/provider-azure/database.azure.crossplane.io/CosmosDBAccount/v1alpha3@v0.17.0#spec-writeConnectionSecretToRef)

Author: Trying to keep the text to <180 characters long per line for readability (except in places where I can't, like the table). I've bumped the "A" in this sentence up to the previous line but left the link on the following line as it's super long.

by Crossplane.

### ACK

**User provided secrets** are provided via a [SecretKeyRef](https://github.com/aws-controllers-k8s/runtime/blob/main/apis/core/v1alpha1/secret.go). 
This allows cross-namespace references to a secret. The specific `Key` of the secret is selected with a `Key string` parameter.

It looks like there may [not currently be support](https://github.com/aws-controllers-k8s/community/issues/700) for key updates/rollover.

**AWS generated secrets** I can't find any examples of. They do not seem to classify "endpoints" as a secret, as shown by
[opensearchservice endpoint](https://aws-controllers-k8s.github.io/community/reference/opensearchservice/v1alpha1/domain/) and
[cluster endpoint](https://aws-controllers-k8s.github.io/community/reference/eks/v1alpha1/cluster/).

## Proposal

### User specified secrets

User specified secrets will be detected and transformed from `string` to a `SecretRef`:

```go
type SecretRef struct {
    // SecretName is the name of the secret. The secret must be in the same namespace as the resource.
    SecretName string

    // Key is the key in the secret to use.
    Key string
}
```

Detection will be done with a combination of:

> 1. Additions to ASO's configuration file to flag particular fields as secret.

Author: Yeah, agreed - that is where I was thinking of putting it. I've made a little note of that here in the doc too just for clarity

2. Using the `"format": "password"` data from Swagger, such as that used by 
   [MySQL Flexible Server](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/mysql/resource-manager/Microsoft.DBforMySQL/stable/2021-05-01/mysql.json#L1662). 
   Note that not all specs have this, for example
   [VMSS does not](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/compute.json#L9397).
3. Using the `x-ms-secret` annotation, which I assume has the same meaning as `"format": "password"` although it's not actually exactly documented anywhere.

This is a place where we can push changes upstream to flag things as passwords if they're not being flagged.

#### Lifecycle
Since these secrets were created _by the user_, the user owns the lifecycle of these secrets. They could be using the same secret across many resources, or 
intending to use this secret on a resource they have not created yet. As such, the lifecycle of these secrets must be controlled by the user.

> Rollover will be supported by triggering events on the associated resource when the secret is modified. Since multiple custom resources might be using 

Author: It will depend on the actual service/API in question. In the case of most databases that I have seen, their administrator account only has a single password, so any time that password is changed for rollover purposes anything using that account would be impacted yes. There's not a lot we can do to fix that though. The DB itself would need to support multiple passwords so that they could roll one and then the other (similar to how `StorageAccount` or `CosmosDB` support multiple keys today).

With that said, I suspect that's not a _huge_ risk. Most places actually using databases in production won't be using the the administrator account for their applications, they would have dedicated accounts for that. If/when we support individual database accounts (a dataplane API) this will come up again but from my reading on the internet the "best practice" seems to be that if your application fails with an auth error to fail out and start again by reading your secret from the secret store - which would be the Kubernetes secret in this case and so while there would be an interruption when rolling the password it wouldn't be a major one - pods would restart, get injected with the new password, and pick back up where they left off. Technically speaking if the password update takes a long time the secret in Kubernetes can be "too new" (as Azure hasn't accepted the new one yet) and if pods restarted early they'd get the wrong secret and fail for a bit. 

That's a really long-winded way of saying: No, I don't think we need to worry about coordination or timing here. If the user asks us to update the password (by changing it in the Kubernetes secret), we should just do it. We aren't going to be able to guarantee 100% uptime unless the service in question supports multiple keys/passwords/secrets, and if the service _does_ support that, we should be able to do what we've designed here and we will automatically get those same 100% uptime guarantees (for the services that support it).

> Rollover will be supported by triggering events on the associated resource when the secret is modified. Since multiple custom resources might be using 

Reviewer: Sounds good. Does any of your commentary here belong in the doc itself?

> Rollover will be supported by triggering events on the associated resource when the secret is modified. Since multiple custom resources might be using 

Author: I've added most of the last paragraph, just for record-keeping

the same secret, this could trigger reconciles on multiple resources. An existing pattern has been established for this in 
[mysqlserver_controller.go](https://github.com/Azure/azure-service-operator/blob/main/controllers/mysqlserver_controller.go#L51) of ASO v1.

#### Open questions

**Should we support KeyVault inputs?**

I think we can get away without it at least initially... but maybe we need to fix up the shape of the `SecretRef` a bit to more easily support it in the future?

```yaml
  ...
  secret:
    kubernetes:
      name: foo
      key: bar
    keyVault:
      url: ??
```

Alternatively, we can keep this focused on Kubernetes secrets and if we want to expand to KeyVault secrets in the future just that _that_ on as a separate property
so that the base (Kubernetes) case is always clean:
```yaml
  secret:
    name: foo
    key: bar
    keyVault:
      url: ??
```

<!-- TODO: Is there specific API guidance from Kubernetes on how to model these unions? There's some stuff discussed here, but it doesn't seem widely adopted: https://github.com/ecordell/enhancements-1/blob/master/keps/sig-api-machinery/20190325-unions.md -->

> Conclusion: TBD

Author: Yeah that's (basically) what Kubernetes suggests I _think_ based on the linked design. I am not actually sure how we'll it's modelled in the backing Go types though. In other REST APIs I've seen the actual SDKs/types end up with an abstract base type with a single property (the `secretType` in this case) and then derived types have other properties. This results in having to perform casts to actually get at the more concrete derived type (at least during GETs).

> Conclusion: TBD

Author: and the usually suggested alternative I've seen for fixing that problem is to instead of doing an inheritance approach do a `oneOf` with a type that looks like:
```
type Example struct {
    optionA *TypeA
    optionB *TypeB
}
```

> Conclusion: TBD

Author: Upon further reading of their spec, actually they suggest:
```go
type Example struct {
    discriminator DiscriminatorEnum
    optionA *TypeA
    optionB *TypeB
}
```

> Conclusion: TBD

Author: I've updated this question with more notes about what Kubernetes recommends. They actually allow for either of the two proposed shapes here, so we will need to decide which one we like more.

      keyDestination:
        kubernetes:
          name: my-secret
        keyVault:
          reference:
            armId: /subscriptions/.../resourceGroups/.../providers/Microsoft.KeyVault/vaults/asokeyvault
          name: my-secret
      # Save the read only keys just to a Kubernetes secret called "my-readonly-secret"
      readOnlyKeyDestination:
        kubernetes:
          name: my-readonly-secret
```

Note that some resources (like CosmosDB `DatabaseAccount`) have multiple kinds of secrets. There might be a `PrimaryKey`, `SecondaryKey`, `ReadOnlyPrimaryKey`, 
and `ReadOnlySecondaryKey`. At the very least we need to support putting the main keys and the readonly keys into two different secrets. To accomplish this,
we can define multiple logical secret groupings in the ASO config and a corresponding `destination` property will be created for each of them. You can see this
done above in the sample.

Some fields such as `Endpoint` (or `AccountName` for other databases) are not _really_ secret, but should be included in these logical secret groups anyway so

> that they're easier for users to inject into pods.

Author: In a lot of ways, yes, it conflicts.
I've added this as an open question so we can discuss further.

> I'm also cautious about classifying things as secretes merely to allow them to be injected into pods.

The issue is that values like `endpoint` are generated by Azure, and there isn't really a way to get those values injected into pods _except_ putting them into a `configMap` or `secret`. We _could_ make a `configMap` with the non-secret values and a `secret` with the secret values but that seems more burdensome/confusing from an operations perspective.

We shouldn't just make the users copy the returned value, as there's no good way to do that in a gitops style workflow.

This is in contrast to _user specified_ values of this sort, which they can just duplicate into 2 places in their `spec`'s - one in the `DB` `AccountName` field (for example) and 1 in a secret of their choice (for use in injecting into their pods if they desire). Since they control this value they can use client tooling/gitops to manage it. It's a bit icky still since the same value is in two places but it's at least doable.

> that they're easier for users to inject into pods.

Reviewer: Fair enough.

get us what we need. Issue [#1978](https://github.com/Azure/azure-service-operator/issues/1978) is tracking this request in more detail.

```go
type ARMDetails struct {
    Endpoint       string
    SubscriptionID string
    Creds          azcore.TokenCredential
    HttpClient     *http.Client
}

type ReconcileDetails struct {
    log                logr.Logger
    recorder           record.EventRecorder
    ARMDetails         *ARMDetails
    KubeClient         *kubeclient.Client
    ResourceResolver   *genruntime.Resolver
    PositiveConditions *conditions.PositiveConditionBuilder
}

type ReconcileOverride interface {
    // BeforeCreateOrUpdate runs before sending the resource to Azure
    BeforeCreateOrUpdate(ctx context.Context, details *ReconcileDetails) (ctrl.Result, error)

    // AfterCreateOrUpdate runs after the resource has been sent to Azure and finished creating, either successfully or with a fatal error.
    // You can determine the current state by examining the Ready condition on obj.
    AfterCreateOrUpdate(ctx context.Context, details *ReconcileDetails) (ctrl.Result, error)

> }

Author: yeah that's probably fair

> }

Author: I've updated the example here to show 2 interfaces.

to the user and expecting them to provide us with the name of a secret that's going to work. We may also need a flag per-secret (or global at the operator 
level?) for if we should purge the secret when we delete. Something like `purgeOnDelete`.

The main downside of a global flag controlling the behavior of purge or delete is that it doesn't give fine-grain control, yet we are proposing to give 
fine-grained control over what vault to put secrets in. This seems like a discrepancy. This seems to suggest we just put this option on every KeyVault reference:

```yaml
spec:
  # Other spec fields elided...
  forOperator:
    secrets:
      # Save the read-write keys (and endpoints) into a kubernetes secret called "my-secret" and a KeyVault secret called "my-secret".
      keyDestination:
        keyVault:
          reference:
            armId: /subscriptions/.../resourceGroups/.../providers/Microsoft.KeyVault/vaults/asokeyvault
          name: my-secret
          purge: false  # Optional: defaults to false
          delete: false # Optional: defaults to true
```

> Conclusion: TBD

Author: The user has deleted the resource the secret corresponds to though - there's no resource left for the secret to correspond to.

Imagine the scenario where the user has a storage account and puts theirs keys into KeyVault. Then they delete their storage account. Does it make sense to leave those keys in KeyVault (and not purge them?). That just means that if you wanted to recreate the storage account you'd need to use a different secret name as the old one is taken for _90 days_ until it is purged.

The keys there have no value though because even if you recreate the exact same named storage account, it'll have different keys.

Still, the proposal here is to by default leave them there (and thus make customers use a different name when recreating), but they can optionally purge them, which is useful especially for cases like testing where you'll probably be creating and deleting similarly named resources and secrets regularly.

[KeyVault secret store CSI driver](https://github.com/Azure/secrets-store-csi-driver-provider-azure) they could import the entire JSON blob and then extract 
the fields they need?

See the [alternate design](#alternative-design) that removes the need for any notion of flattening, which also resolves this.

Conclusion: TBD

**With the proposed design, there is no static understanding of the fields that will be present in the secret**

The fact that you will get 3 fields (`primaryKey`, `secondaryKey`, `endpoint`) in your CosmosDB `DatabaseAccount` secret is totally opaque to users.
Do we just need to document this? See proposal below for a possible alternative that solves this discovery problem.

Conclusion: TBD

**How do we ensure that we aren't overwriting secrets that we don't own?**

The operator must support updating the secret as keys can change or be rolled over. On the other hand we should present a clear error if the user has 
accidentally pointed two resources at the same secret. The first resource should be unaffected and work normally while the second resource should encounter

> a reconcile error stating that the secret in question already exists.

Author: The _very_ first writer wins - once a resource has created a secret and marked itself as owning it (using Kubernetes ownership or the proposed tag), all subsequent attempts by any resource other than that original winner will error.

> a reconcile error stating that the secret in question already exists.

Author: Basically the original writer locks the secret to itself

   cleaner to store it in `Status` but technically that can be lost, whereas `Annotations` won't be. This should work for both KeyVault and k8s secrets and 
   while it's a bit icky I think gives the best experience.

Conclusion: TBD

**Should we drop the `Destination` suffix in the field names?**

For example, the proposal for CosmosDB `DatabaseAccount` was:
```yaml
spec:
  # Other spec fields elided...
  forOperator:
    secrets:
      keyDestination:
        # stuff elided...
      readOnlyKeyDestination:
        # stuff elided...
```

```yaml
spec:
  # Other spec fields elided...
  forOperator:
    secrets:
      key:
        # stuff elided...
      readOnlyKey:
        # stuff elided...
```

Conclusion: TBD

**Do we allow writing a secret to a namespace where the resource isn't?**

No, see: https://github.com/kubernetes/community/pull/5455

Conclusion: TBD

**Do we support writing multiple destinations at the same time?**

If/when we add support for writing secrets to KeyVault rather than to Kubernetes secrets, would we also enable support for writing the same secret to _both_

> KeyVault and Kubernetes at the same time?

Author: At the same time in this context meant concurrently. If one or the other (or both) fails, we would return a reconciler error and attempt to re-reconcile (the same as if creating a resource in Azure failed, we keep trying and keep driving towards goal state).

> KeyVault and Kubernetes at the same time?

Author: I've removed the "at the same time" language here, just because really my question is more about "do we support writing to both". Have also added an example `spec` highlighting what I am talking about here.

I think we should.

Conclusion: TBD

**What client should we use for issuing the ListKeys/GetKeys request?**

We don't automatically generate methods for performing these calls. It seems easiest to just use the corresponding Azure SDK for this.
This does mean that we're taking a dependency on the SDK where we didn't have one before, but the overall footprint of usage is pretty small

> and it's going to be easier to do that than it is to write the methods ourselves.

Author: No you actually get them all by default, the golang SDK ships all versions all the time, and you can import them all side by side (obviously you'll need to call one `compute2021` and one `compute2022` but other than that no issues.

Conclusion: TBD

**What API version should we use for issuing the ListKeys/GetKeys request?**

There are a few options here. The custom hooks could be applied to either the customer facing resource version _or_ to the internal storage version.

> Applying the hooks to the storage version is easier, but means that the same hook will be run regardless of which customer 

Author: Yep, as discussed yesterday my preference would be that this uses the same "hooks" mechanism as other customization we have.

   (and might not even know about).
2. We can make a failing `ListKeys` (or equivalent) call fatal. Many services support ways lock down or forbid access to the `ListKeys` (or equivalent) API.
   This allows users to block access if they would like to require AAD authentication with their resource. As long as the operator is not calling those APIs
   unless asked (in the `Spec` of the resource) we can make their failure fatal. This gives users an easy way to drive towards full AAD usage in ASO and elsewhere.

We'll dig more into how we might accomplish this a bit later.

### Work well with GitOps

Resources should strive to play well with GitOps. This means that when a new resource is deployed it is capable of adopting an already existing ARM resource
if one exists. This also must apply to secrets. As mentioned above one of the reasons that 
[we do not want to be generating secrets on the users behalf](#aso-should-not-be-generating-secrets-on-the-users-behalf) is that it makes redeploy
and resource adoption hard because the user has no configuration representing the generated secret.

## Kinds of secrets

> There are two main types of secrets we need to consider.

Reviewer: ```suggestion
There are two main types of secrets we need to consider, differing primarily by the origin of the secret.
``

**User provided secrets**: Secrets provided by the user at resource creation time.

> **Azure generated secrets**: Secrets created by Azure, and returned in a special `GetMeTheSecrets` API call.

Reviewer: ```suggestion
**Azure generated secrets**: Secrets created by Azure, and returned by a special `GetMeTheSecrets` API call.
```

```go
type SecretReference struct {
    // Name is the name of the secret. The secret must be in the same namespace as the resource.
    Name string

    // Key is the key in the secret to use.
    Key string
}
```

Detection will be done with a combination of:
1. Additions to ASO's configuration file to flag particular fields as secret (probably in the `ObjectModelConfiguration` section).
2. Using the `"format": "password"` data from Swagger, such as that used by 
   [MySQL Flexible Server](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/mysql/resource-manager/Microsoft.DBforMySQL/stable/2021-05-01/mysql.json#L1662). 
   Note that not all specs have this, for example
   [VMSS does not](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/compute/resource-manager/Microsoft.Compute/stable/2020-12-01/compute.json#L9397).
3. Using the `x-ms-secret` annotation, which I assume has the same meaning as `"format": "password"` although it's not actually exactly documented anywhere.

This is a place where we can push changes upstream to flag things as passwords if they're not being flagged.

#### Lifecycle

> Since these secrets were created _by the user_, the user owns the lifecycle of these secrets. They could be using the same secret across many resources, or 

Reviewer: I think this should be present tense.

```suggestion
Since these secrets arecreated _by the user_, the user owns the lifecycle of these secrets. They could be using the same secret across many resources, or 
```

for now. In any case it's not totally clear to me that we really need the value that it is adding, and it seems to add a significant amount of update complexity.

We need to decide which of the above shapes we like more, and also if we do or do not want a `secretType` discriminator.

**Conclusion:** Not needed for v1 of the feature, but we need to have a shape ready that makes sense for when we do.

```yaml
  ...
  secret:
    type: Secret # Or KeyVault
    name: foo
    key: bar
    keyVaultReference:  # This is the standard Reference type we use elsewhere
      armId: ...
```

**Is an `AccountName` a secret?**

If the service is returning it in the resource `GET`, then strictly speaking it is not a secret. Since the field isn't secret, we will not transform it
to be a `SecretReference`. This is for two reasons:
1. We won't be able to automate it, since as far as the OpenAPI specification is concerned it isn't a secret.
2. The primary reason to classify this as a secret would be so that users could then inject the value from their secret into a pod. There is a workaround

>    for this though since the user could (using Kustomize or something) do this already I think.

Reviewer: ```suggestion
   for this though since the user could (using Kustomize or similar) do this already.
```

There were some requests for this in ASOv1, see [this](https://github.com/Azure/azure-service-operator/issues/1396) for example.

This has security implications, so initially at least the answer should be no. See: https://github.com/kubernetes/community/pull/5455.

**Conclusion:** No

**What's the difference between `x-ms-secret` and `format: password`?**

Unclear currently. I have a question out to the Swagger team. Current plan is to just treat them the same.

**Conclusion:** They are not the same, `x-ms-secret` implies that the secret is not returned in the `GET`, whereas `format: password` is just an annotation that
as far as we can tell is not actually consumed by anything. For our case, we can treat them the same.

### Azure generated secrets

Azure generated secrets will be _optionally_ downloaded to a Kubernetes or KeyVault secret. Users instruct the operator to download
the secrets associated with a resource by supplying a `SecretDestination` in the `Spec`.

The optionality of this step is key for [preferring managed identity](#prefer-managed-identity), as it allows those in AAD/Managed identity cases to avoid

> secrets they don't want/won't use being put into their namespace.

Reviewer: ```suggestion
secrets they don't want/won't use being retrieved into their namespace.
```

```go
// Note: This is the same type that is used for specifying references to user created secrets/
// reproduced here for clarity
type SecretReference struct {
	Name string
	Key string
}

// Using a different type for the destination as there are some things that are destination specific (such as adding annotations or labels, 
// which don't make sense on )
type SecretDestination struct {
	SecretReference
}
```

#### Example: Explicit secrets w/ Azure Storage (Simple)

```yaml
spec:
  # Other spec fields elided...
  operatorSpec:
    secrets:
      primaryKey:
        name: my-secret
        key: PRIMARY_KEY
      secondaryKey:
        name: my-secret
        key: SECONDARY_KEY
      endpoint:
        name: my-secret
        key: ENDPOINT
```

#### Example: Explicit secrets w/ CosmosDB and multiple secret destinations (Complex)

Here is what a more complex resource might look like if we also supported `KeyVault` as a secret type.

> **Note**: Not planning to support `KeyVault` initially.

Reviewer: ```suggestion
**Note**: We are not planning to support `KeyVault` initially.
```

        key: PRIMARY_KEY
     readOnlySecondaryKey:
        type: Secret
        name: my-readonly-secret
        key: SECONDARY_KEY
     endpoint:
        type: Secret
        name: my-endpoint
        key: ENDPOINT
```

Then if they want to use the secondary key for a pod they would mount the endpoint and the primary/secondary keys from
`my-readonly-secret`.

While this isn't quite as clean as it could otherwise be by supporting multiple secret destinations for each
secret kind (`endpoint`, `readOnlyPrimaryKey`, etc) it's simpler to implement and has fewer failure modes as we don't
have to deal with the possibility of the user specifying 10 destinations where we wrote some but failed to write others.

**Conclusion:** No, we will not support this.

**What client should we use for issuing the ListKeys/GetKeys request?**

We don't automatically generate methods for performing these calls. It seems easiest to just use the corresponding Azure SDK for this.
This does mean that we're taking a dependency on the SDK where we didn't have one before, but the overall footprint of usage is pretty small
and it's going to be easier to do that than it is to write the methods ourselves.

> Conclusion: TBD

Reviewer: ```suggestion
**Conclusion**: TBD
```

Author: Good eye

Author: Actually we decided on this too. Will use the Azure SDK for it.

5. It doesn't work well with GitOps. A number of customers have expressed the desire to move resources between namespaces. In theory this is easy - 
   just create the exact same resources in a different namespace (pointing at the same Azure resources), mark the old resources as `skip-deletion` (so that
   the backing Azure resources are not deleted), and delete the original namespace. If the definition of the secret isn't part of the users GitOps flow this 
   becomes more difficult as the first namespace has  the secret (which was created by ASO) and that secret must be cloned manually to the second namespace,
   otherwise the creation of the resource in the second namespace will attempt to generate a new secret.

We could investigate allowing both automatically generated secrets (the default) and user specified secrets (opt-in). That avenue results in the operator 
shouldering all of the complexity burden of both though. 

Given ASO's approach as a low level toolkit providing the ability to create Azure resources, at least 
at this time we should avoid the added complexity of generating user secrets.

> ### Prefer Managed Identity

Reviewer 2: Why do want to give users an option of not using managed identity? Are we just limited by resources supporting managed identity or is there customer value in being able to choose?

Author: There are "good reasons" to not use managed identity. For example we use service principal for our ASO test environment, because managed identity requires you to run in Azure, which is annoying when you could otherwise kick the tires/test outside of Azure.

There are also some resources that don't support managed identity, or more often don't support it for all scenarios.

but I think it's _mostly_ that the service itself doesn't force you to use managed identity. If we for example forced all users of Azure storage to use managed identity, it would preclude users who just want to lift+shift their stuff into ASO and have it adopt their existing resources.

With that said, this preference for managed identity here means we shouldn't default to pulling secrets and we should probably (as best we can) document with managed identity first.

I think we should.

Conclusion: TBD

**What client should we use for issuing the ListKeys/GetKeys request?**

We don't automatically generate methods for performing these calls. It seems easiest to just use the corresponding Azure SDK for this.
This does mean that we're taking a dependency on the SDK where we didn't have one before, but the overall footprint of usage is pretty small
and it's going to be easier to do that than it is to write the methods ourselves.

Conclusion: TBD

**What API version should we use for issuing the ListKeys/GetKeys request?**

There are a few options here. The custom hooks could be applied to either the customer facing resource version _or_ to the internal storage version.
Applying the hooks to the storage version is easier, but means that the same hook will be run regardless of which customer 

#### Alternative design

Would we prefer an alternative design where each individual field that goes into a secret is called out explicitly? It's noisier but more flexible and also 
probably more discoverable, at least in theory? Here's what it might look like for CosmosDB `DatabaseAccount`

```yaml
  # Other spec fields elided...
  forOperator:
    secrets:

>       primaryKeyDestination:

Reviewer 3: Maybe this should be a list with names rather than a map, for better compliance with the k8s guidelines? I guess we still have a lot of maps that we inherit from ARM, but it's probably worth not making it worse.

Author: My intent at least was that it is an object, not an open ended map.

included in secrets we write. That will include things like `endpoint`. This means that `endpoint` for a SQL Server would show up in two places,
in the status _and_ possibly in the secret written by the operator (assuming that the user has instructed us to write the endpoint someplace).

The main reason for writing this into two places is:
1. We already have it in the `Status` and it doesn't make sense to remove it as it's part of the resource payload from Azure.
2. We want the ability to put it into a Secret so that if users want to inject it into their pods as an environment variable it's easy to do so.
   There is no good way to inject environment variables from CRD Status's.

**Short-lived "tokens"**: Like Storage SAS.

We will not support these sorts of secrets.

### A special note on KeyVault
We have an [open issue](https://github.com/Azure/azure-service-operator/issues/1894) asking for support to manage KeyVault secrets via ASO.

We have to be _very_ careful about support for KeyVault secrets (or certs, keys, etc), as the whole point of KeyVault is as a secure place to store your 
secrets.

> If you're instead creating those secrets via the operator then you have by definition also located the secret in Kubernetes, which somewhat defeats the purpose 

Reviewer 2: Good point. Do we then create a KeyVault operator to essentially do the role of the secret store CSI driver, or is there anything else we can add?

Author: I don't know what the right story is here honestly. It kinda feels like, if you want keyvault _for security reasons_, then you by definition do not want your secrets in Kubernetes and so no Kubernetes based solution for storing (or creating) the secrets can work for you.

If you want KeyVault because Azure "makes" you for some case or another, then you might not care and an ASO KeyVault Key resource that takes a k8s `Secret` as input might make sense for that case. We'd need to be sure that case was common enough that we care about it though, and I am not 100% sure it is.

Note: This doesn't stop us from putting the secrets we get from Azure into KeyVault (which this proposal suggests support for). That's fine because we can bypass Kubernetes secrets entirely and then the user could use the KeyVault CSI driver to inject those same secrets that ASO created in KeyVault (via `ListKeys` on storage for example) into their pod.
